pandas/core/frame.py

```diff
@@ -269,7 +269,10 @@
 ----------%s
 right : DataFrame or named Series
     Object to merge with.
-how : {'left', 'right', 'outer', 'inner', 'cross'}, default 'inner'
+how : {'left', 'right', 'outer', 'inner', 'cross'}
+    default if `condition` not defined: 'inner'
```

Comment:
why would the default change?

Reply:
my assumption is that the vast majority of the time (if not 100% of the time) that someone uses a conditional join, they want only the condition to define the merge, which can only be achieved by an initial cross join. Feels clunky to me to have to explicitly state `how='cross'` when you are already making the decision to pass a custom merge condition.

Reply:
i don't buy this, why is a conditional merge always 'cross'? can you show an example?

Reply:
@jreback - Sorry for the incoming wall of text. I am moving the goal posts a bit, and I'm not sure if some of the stuff I'm introducing below has been considered and subsequently dismissed in the past, also if this is out of scope for the library, please let me know as well.

cc @mroeschke as well since you were involved in some of the conditional join discussions

First, I think there's 2 things to address here:

1) should we modify the default
2) should we allow other types of joins with condition besides cross

My thoughts are:

1) I don't feel strongly enough that I would push hard for this. I would be ok leaving it as inner.
2) To your question of "why is a conditional merge always cross?". Conceptually, I don't think it needs to be. However, considering our initial chunked approach, I think that only a cross (and actually inner too) join can preserve the same behavior from initial to final desired implementation.

Before I continue on why I think other join types wouldn't stay consistent, I thought about the API a bit more, and I think that actually it makes more sense for the condition to be provided through the `on` keyword (this is relevant to my point about preserving behavior). This is more consistent with how joins work in general. The fact that in the current api we provide e.g. `on=[A, B]` (ignoring the other left/right_on/index args for now) is really just a shorthand for `on=left.A == right.A and left.B == right.B`. This is a specialization of the more general conditions of what constitutes a match in a join. So I think it would make sense to provide your callable condition in the `on` argument (as an addition to what is allowed in `on`, all the existing functionality would remain).

I actually wouldn't recommend anymore a separate `condition` argument (especially not in tandem with `on`) because essentially with the way `on` works right now it will act as an `AND col==col` addition to your condition which is overly restricting. 

Getting to why other join types won't stay consistent with the final approach now. As it stands right now the fact that we are considering a chunked approach to start with, this will also enforce some limitation on what we can do. With a simple naive chunked approach, you can't recreate a left, right or outer join by joining the cartesian product pairs of chunks. This will introduce duplicate record artifacts into your result. Even if you could do the left/right/outer joins in the chunked manner you'd need to satisfy the SLA of "preserving the left or right's key order", which will not be trivial/performant after aggregating the chunked results.

Cross/Inner joins where we can provide a initial chunked implementation that will satisfy the api of and generalizable nature of a final solution. As of today we could very easily setup the interface to work like so, without interfering with any existing functionality (same mechanism as current PR, just allow `on` to be callable and remove `condition`):

```
left.merge(right, how='cross', on=lambda dfx: dfx.A_left == dfx.A_right)
```

For inner join with `on=callable`, none of the existing merging mechanisms would be used, it would actually be identical to cross join with on=callable. 

All this said the desired final state would require quite a bit of additions to the cython joining code: `pandas/_libs/join.pyx`. All the existing implementations would still exist for when you use merge with the classic `on=<some column names>`. Since it looks like the existing cython join code uses factorized column values and is optimized for equality,  we'd need to write new (separate/additional) versions of those functions that can compare the actual values to determine when a join condition is met.

Reply:
I take this back.
See here: https://github.com/pandas-dev/pandas/pull/42964#issuecomment-915423616